### ISSUE TITLE
In consistent behavior of existence check for empty folders

### ISSUE THREAD

**Issue by drorata** [dask/s3fs#156]
Consider the following minimal example:

```python
import moto
import boto3
import s3fs

with moto.mock_s3():
    print("Case 1")
    conn = boto3.client('s3', 'us-west-2')
    conn.create_bucket(Bucket='test-bucket')
    
    s3 = s3fs.S3FileSystem()
    
    s3.mkdir('test-bucket/foo/')
    print(s3.ls('test-bucket'))
    print(s3.exists('test-bucket/foo/'))
    s3.touch('test-bucket/foo/bar')
    print(s3.exists('test-bucket/foo/'))
    
with moto.mock_s3():
    print("Case 2")
    conn = boto3.client('s3', 'us-west-2')
    conn.create_bucket(Bucket='test-bucket')
    
    s3 = s3fs.S3FileSystem()
    
    s3.mkdir('test-bucket/foo')  # A missing trailing / is the only difference
    print(s3.ls('test-bucket'))
    print(s3.exists('test-bucket/foo/'))
    s3.touch('test-bucket/foo/bar')
    print(s3.exists('test-bucket/foo/'))
```

For which the output is:

```
Case 1
['test-bucket/foo']
False
True
Case 2
['test-bucket/foo']
True
True
```

Note that in the first case, as long as the folder `foo` is empty it can be listed but not detected by `s3.exists`. In contrast, in the second case, when the trailing `/` is removed when creating the directory, `.exists` detects the folder even when it is empty.

I also created empty folders directly from the S3 console, and still `.exists` couldn't find them.

If find it an un-expected behavior. Is it a bug or a feature?

**Comment by drorata** [comment]
Related, and somehow confusing as well, compare `s3.mkdir('my-bucket/foo')` and `s3.mkdir('my-bucket/bar/'). The former creates an empty file and the latter a directory.

**Comment by martindurant** [comment]
s3 doesn't actually have directories, only files. mkdir probably shouldn't do anything, ever. I'm not sure what making a directory in the console really does.

**Comment by drorata** [comment]
It is not 100% correct. 

Consider:

```python
s3.mkdir('my-bucket/testing/with_trailing_slash/')
s3.mkdir('my-bucket/testing/without_trailing_slash')
```

Yields the following:

<img width="1531" alt="screenshot 2018-11-30 at 14 38 20" src="https://user-images.githubusercontent.com/2905978/49292477-a20bc500-f4ad-11e8-9011-b2ecc2a6e226.png">

One is treated as a folder/directory and the other as a file.

**Comment by martindurant** [comment]
I understand that there is also a bug in s3, and I'll look into it (or you are welcome to!).

My point is, perhaps mkdir should simply error when given anything longer than a bucket, since the real way to make a folder is to make a file within it.

**Comment by drorata** [comment]
I tried to replace `mkdir` with:

```python
def mkdir(self, path, acl="", **kwargs):
        """ Make new bucket or empty key

        Parameters
        ----------
        acl: str
            ACL to set when creating
        region_name : str
            region in which the bucket should be created
        """
        acl = acl or self.s3_additional_kwargs.get('ACL', '')
        path = path.rstrip('/') + '/'
        self.touch(path, acl=acl, **kwargs)
```

But some tests failed and I am not sure why...

**Comment by drorata** [comment]
BTW, @martindurant , could it be it is related also to https://stackoverflow.com/q/53558319/671013 ?

**Comment by martindurant** [comment]
I am not a fan of using `touch` as you suggest, because this actually creates an empty file on the server.
It would be interesting to know what information is returned on a "directory" that you create via the web interface.

**Comment by drorata** [comment]
If I create `foo` in the bucket's root (`my-bucket`), then `S3FileSystem().ls('my-bucket')` includes `foo` in the returned list. However, both `S3FileSystem().ls('my-bucket/foo')` and `S3FileSystem().ls('my-bucket/foo/')` raise a `FileNotFoundError.

**Comment by drorata** [comment]
I'm not sure what you don't like about `touch`. It is used by `mkdir` and it is an expected behavior that touching a non-existing file creates an empty one. In the S3 case, I guess depending on trailing `/` it is determined whether a new empty file or directory is created.

**Comment by martindurant** [comment]
Indeed, `mkdir` currently calls `touch`, the question is whether it should. It would surprise most people that, when asking for the creation of a directory, you actually get a (empty) file. Of course, the remote system doesn't really have directories.

I do not seem to find the same as your test case on real S3:
```python
>>> s3.mkdir('MDtemp/foo')
>>> s32.ls('MDtemp/foo')  # other instance
['MDtemp/foo']
>>> s32.ls('MDtemp/foo/')
['MDtemp/foo']
>>> s32.exists('MDtemp/foo')
True
```

Question: why do you find you want to make directories as all?

**Comment by drorata** [comment]
As far as the S3 console looks like, it seems like there is a difference between directories and files. In particular, you can have an empty directory. As far as I understand, this is the case when an object's name ends with a `/`.

I implement a logic where first I want to have a directory (and verify its name, location etc.) and only then add content to it using different methods.

I didn't understand the example you shared. Can you elaborate?

**Comment by martindurant** [comment]
The example was, I think , what you were trying to do.

From your description, I finally understand what you mean, and I agree that your solution of simply ensuring the path ends with '/' is reasonable. If you submit the PR, we can figure out why things are failing.

**Comment by drorata** [comment]
See PR I started.

**Comment by dvukolov** [comment]
I am experiencing several issues, some of which may be related to the new code. Please forgive me if do not understand how s3fs operates, what is expected and what is not — the result is at times confusing. Here is a test case:

```python
import s3fs
s3 = s3fs.S3FileSystem()
bucket = 'test-bucket'

s3.mkdir(bucket)
assert bucket in s3.ls('')  # ok
s3.ls(bucket)  # raises a FileNotFoundError

s3.mkdir(f'{bucket}/dir1')
s3.ls(bucket)    # still raises a FileNotFoundError

s3.touch(f'{bucket}/file')
assert f'{bucket}/file' in s3.ls(bucket)  # ok
assert f'{bucket}/dir1' in s3.ls(bucket)  # ok

s3.mkdir(f'{bucket}/dir2')
assert f'{bucket}/dir2' in s3.ls(bucket)  # raises AssertionError
assert f'{bucket}/dir2' in s3.ls(bucket, refresh=True)  # ok

s3.ls(f'{bucket}', detail=True)  # ok
# directories have StorageClass 'DIRECTORY' and names without a trailing slash

s3.ls(f'{bucket}/dir1')  # seemingly not ok. should it return an empty list instead?
# returns ['test-bucket/dir1/']

s3.ls(f'{bucket}/dir1', detail=True)  # seemingly not ok
# the directory has StorageClass 'STANDARD' and a name with a trailing slash

s3.mkdir(f'{bucket}/dir1/dir3')
assert f'{bucket}/dir1/dir3' in s3.ls(f'{bucket}/dir1')  # raises AssertionError
assert f'{bucket}/dir1/dir3' in s3.ls(f'{bucket}/dir1', refresh=True)  # ok

s3.ls(f'{bucket}/dir1')  # seemingly not ok. should it return just dir3?
# returns ['test-bucket/dir1/', 'test-bucket/dir1/dir3']
```

**Comment by martindurant** [comment]
Again, it's not entirely clear what the behaviour *should* be, since S3 simulates folders but it really a key-value store. Perhaps s3fs simply shouldn't attempt `mkdir`... Similarly, it is not clear whether `ls(f'{bucket}/dir1')` should return `'test-bucket/dir1/'` or not, since now a key with exactly that value does exist. Note that you can perfectly well have a file and a directory-key and also a prefix all with identical keys.

From your code above, it does appear that the current implementation is not calling `invalidate_cache` at the right moment.

**Comment by dvukolov** [comment]
Martin, might it make sense to mimic AWS CLI behaviour? In case of listing an empty directory, it returns nothing.

**Comment by dvukolov** [comment]
And listing a directory with contents, returns only the contents, excluding the directory itself.

**Comment by martindurant** [comment]
@dvukolov , if you can specify exactly the behaviour you expect in every case, and (even better) phrase it in API calls and make a test case, please do! Getting the logic right here, including when we decide that a directory is missing as opposed to empty and when to clear the cache is very tricky!

**Comment by dvukolov** [comment]
There is a related outstanding issue apache/arrow#3286 with isfile() and isdir() in pyarrow.filesystem.S3FSWrapper. Is there a reason why these methods are implemented in the wrapper class and not in s3fs.S3FileSystem?

**Comment by martindurant** [comment]
> Is there a reason why these methods are implemented in the wrapper class and not in s3fs.S3FileSystem?

I suppose they were not interested in pushing a PR upstream?

Note that [fsspec](https://github.com/martindurant/filesystem_spec/blob/master/fsspec/spec.py#L400) has both, and s3fs will derive from that spec at some point. Also, since an s3 path can be simultaneously a key and  a prefix, it is possible for `isdir(path)` and `isfile(path)` to be simultaneously true.
